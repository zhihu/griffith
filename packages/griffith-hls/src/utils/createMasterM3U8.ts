type Item = {
  bandwidth: number
  resolution: {
    height: number
    width: number
  }
  source: string
}

export default function createMasterM3U8(list: Item[]) {
  const result = []
  result.push('#EXTM3U')

  list.forEach((item) => {
    const meta: Record<string, unknown> = {
      'PROGRAM-ID': '1',
      BANDWIDTH: String(item.bandwidth),
    }
    if (item.resolution) {
      const {width, height} = item.resolution
      meta.RESOLUTION = `${width}x${height}`
    }
    result.push(
      `#EXT-X-STREAM-INF:${Object.entries(meta)
        .map(([key, value]) => `${key}=${value as string}`)
        .join(',')}`
    )

    result.push(item.source)
  })

  return result.join('\n')
}
