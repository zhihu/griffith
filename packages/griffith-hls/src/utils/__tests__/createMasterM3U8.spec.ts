import createMasterM3U8 from '../createMasterM3U8'

describe('create master m3u8', () => {
  it('list with resolution', () => {
    const list = [
      {
        bandwidth: 1029120,
        resolution: {
          height: 478,
          width: 848,
        },
        source:
          'http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/sd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8',
      },
      {
        bandwidth: 2463744,
        resolution: {
          height: 720,
          width: 1280,
        },
        source:
          'http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/hd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8',
      },
    ]

    expect(createMasterM3U8(list)).toMatchSnapshot()
  })

  it('palylist without resolution', () => {
    const list = [
      {
        bandwidth: 1029120,
        source:
          'http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/sd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8',
      },
      {
        bandwidth: 2463744,
        source:
          'http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/hd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8',
      },
    ]
    expect(createMasterM3U8(list)).toMatchSnapshot()
  })
})
