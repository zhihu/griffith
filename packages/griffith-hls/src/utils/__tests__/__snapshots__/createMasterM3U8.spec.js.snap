// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create master m3u8 list with resolution 1`] = `
"#EXTM3U
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1029120,RESOLUTION=848x478
http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/sd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2463744,RESOLUTION=1280x720
http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/hd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8"
`;

exports[`create master m3u8 palylist without resolution 1`] = `
"#EXTM3U
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1029120
http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/sd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2463744
http://zhihu-video-output.oss-cn-hangzhou.aliyuncs.com/test/hd-m3u8/999f95fc-0346-11e9-b494-0a580a44d740.m3u8"
`;
